typing/typedecl.ml

```diff
@@ -657,16 +657,17 @@ let check_well_founded env loc path to_check visited ty0 =
     if TypeSet.mem ty parents then begin
       (*Format.eprintf "@[%a@]@." Printtyp.raw_type_expr ty;*)
       let err =
-        let reaching_path =
+        let reaching_path, rec_abbrev =
           (* The reaching trace is accumulated in reverse order, we
              reverse it to get a reaching path. *)
           match trace with
           | [] -> assert false
-          | trace ->
-              List.rev trace in
-        if match get_desc ty0 with
-          | Tconstr (p, _, _) -> Path.same p path
-          | _ -> false
+          | Expands_to (ty1, _) :: trace when (match get_desc ty1 with
+              Tconstr (p,_,_) -> Path.same p path | _ -> false) ->
+                List.rev trace, true
+          | trace -> List.rev trace, false
```

Comment:
You are changing the definition of `rec_abbrev` in a non-trivial way here. Before we would check if `ty0` starts with a `path` constructor, which is a static condition (`ty0` and `path` never change during the computation) purely depending on the callsite. Now you check if the last type before the cycle is found (not the type that is actually the proof of circularity, but the one before) is a `path` constructor. What is the intuition here, beyond the fact that it looks nice in the tests?

It would be easier for me to think about this change if the first commit correctly committed its test state. I have the impression that the first commit contains the old testsuite results from before you rebased, and then this commit contains the update to the testsuite corresponding to both commits. If the first commit correctly updated the testsuite, then in this one commit we would see the test diff that corresponds to your change here, which is currently not visible anywhere -- I have to guess.

Reply:
For the tests, the problem is that one first has to bootstrap before running them, so you don't get the originals.
I could have added an intermediate commit with the results for the first commit, but they were actually very different from what we had before, in a degraded way, which is why I did further changes first.

Now, for the rationale behind `rec_abbrev`.
The point is that, for an abbreviation, due to the way caching works, we only get the same node when we expand it.
So if you find a cycle, you should check the path before expansion. Before this was done by updating `ty0`, and didn't work very well. But thanks to your trace it is now easy to get the node before expansion.

Reply:
Why do we need a bootstrap for this PR? Are the types stored in .cmi modified by the fact that we expanded abbreviations differently during this check?

Reply:
(I still have the impression that the testsuite update could be synchronized with the code update for each commit, possibly by doing a bootstrap before running the test and then not committing the bootstrap. At least I would see in each commit what the error messages generated by the new code is.)

Reply:
Thanks for the comment on the expansion. It may be worth having some comment in the code to explain this.

Reply:
Now that I'm convinced by your explanation, I care less about the commit structure for the testsuite update. I would still be curious to know why we need a bootstrap, and welcome a code comment on the `rec_abbrev` stuff.

(Note: if the trace contains a `Contains` construct after the expansion, your code will have `rec_abbrev = false`, do we want to look through the trailing `Contains` moves for the last expansion in this case?)

Reply:
I don't think one needs to look through the trailing `Contains`: the point is that you will get a cycle as soon as you expand an inner occurrence of the defined type. So you get the cycle before going through an extra `Contains`.